### PR TITLE
V3.0.0 - replace superagent with axios

### PR DESCRIPTION
**Pull request by max-degterev**
Breaking changes:
 - Replaced `superagent` with `axios`
 - Replaced `api` with `baseUrl` in global config
 - Instead of returning either object with `request` and `promise` network now always returns promise.
 - Request object itself is inaccessible, to cancel requests use `getAbortCallback` which receives an abort callback as first argument.
 - `createApi` is removed. Actions now get access to the `createRequest` function instead of `createApi`. Only difference - you need to provide `token` manually to every call.
 - Store now has proper `isNetworkActive` state that sets to false only when all requests are done. Request actions now have `isRequestActive` flag in their meta (was isNetworkActive). Renamed network store state `activeCount` => `activeRequestsCount`.

New:
 - Added global `requestHook` and `responseHook`

### CODE REVIEW

**Comment by max-degterev** [comment]
TODO:
 - ~Add original config to the request and response hooks~
 - ~Fix querystring array handling - force use default `stringify` from `querystring`~

File: src/network/configuration.es [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
+  baseUrl = options.baseUrl;
   regex = new RegExp(`^https:\\/\\/(?:[\\w-]+\\.)?${escapeStringRegexp(options.trustedDomain)}`);
+  requestHook = options.requestHook;
+  responseHook = options.responseHook;
 };
 
-export const getApiRoot = () => api;
+export const getBaseUrl = () => baseUrl;
 export const getTrustedDomainRegex = () => regex;
+export const getGlobalHooks = () => ({ requestHook, responseHook });
```
Comment (atomiomi) [inline src/network/configuration.es]: why do we need this?

File: src/network/middleware.es [hunk 2]
```diff
@@ -14,43 +13,37 @@ export default (store) => (
...
-      let result;
       if (typeof request === 'function') {
-        result = request(createApi(state.token), store.dispatch, state);
-
-        if (!result || !result.request || !result.promise) {
-          throw new Error('request function must return a hash with `request` and `promise` props');
-        }
-
-        promise = result.promise;
+        promise = request(createRequest, store.dispatch, state);
```
Comment (atomiomi) [inline src/network/middleware.es]: why pass it? we can then import it in actions module

File: src/network/middleware.es [hunk 3]
```diff
@@ -14,43 +13,37 @@ export default (store) => (
...
+        if (!promise) throw new Error('request function must return a promise');
```
Comment (atomiomi) [inline src/network/middleware.es]: `if (!promise || !promise.then)`?

File: src/network/reducer.es [hunk 4]
```diff
@@ -1,11 +1,11 @@
...
-  if (action.meta && typeof action.meta.isNetworkActive === 'boolean') {
-    const isNetworkActive = action.meta.isNetworkActive;
-    let activeCount = state.activeCount || 0;
-    if (isNetworkActive) activeCount += 1;
-    else activeCount -= 1;
+  if (action.meta && typeof action.meta.isRequestActive === 'boolean') {
+    const diff = action.meta.isRequestActive ? 1 : -1;
 
-    return { activeCount, isNetworkActive };
+    const activeRequestsCount = Math.max((state.activeRequestsCount || 0) + diff, 0);
```
Comment (atomiomi) [inline src/network/reducer.es:5]: by using `Math.max` `activeRequestsCount` should never be less than 0

File: src/network/middleware.es (hunk 2 above)
Comment (max-degterev) [reply src/network/middleware.es]: Yea could prob remove this - not used

File: src/network/reducer.es (hunk 4 above)
Comment (max-degterev) [reply src/network/reducer.es:5]: Correct, negative requests make no sense. If it happens something is broken.

**Pull request merged**